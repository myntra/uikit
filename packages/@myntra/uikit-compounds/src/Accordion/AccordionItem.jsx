import React from 'react'
import PropTypes from 'prop-types'
import { Transition } from 'react-transition-group'
import { classnames } from '@myntra/uikit-utils'

import styles from './Accordion.css'

/**
 @since 0.3.0
 @status EXPERIMENTAL
 @example
 <Accordion.Item title={<span>Collapsed</span>}>
  <span>Expanded</span>
 </Accordion.Item>
*/
function AccordionItem({ show, children, title, onClick, ...props }) {
  return (
    <div {...props}>
      {React.cloneElement(title, { onClick })}
      <Transition in={show} timeout={350}>
        {state => {
          return (
            state !== 'exited' && (
              <div
                className={classnames({
                  collapse: state !== 'exiting',
                  collapsing: state === 'exiting',
                  show
                }).use(styles)}
              >
                <div>{children}</div>
              </div>
            )
          )
        }}
      </Transition>
    </div>
  )
}

AccordionItem.propTypes = {
  /** @private */
  className: PropTypes.string,
  /** @private */
  onClick: PropTypes.func,
  /** @private */
  show: PropTypes.bool,
  children: PropTypes.any.isRequired,
  title: PropTypes.node.isRequired
}

export default AccordionItem
