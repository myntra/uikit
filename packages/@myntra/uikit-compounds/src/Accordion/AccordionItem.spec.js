import React from 'react'
import { mount } from 'enzyme'
import AccordionItem from './AccordionItem'

function prepareForTest(props) {
  return mount(
    <AccordionItem {...props} title={<div className="title">Title</div>}>
      <div className="body">Body</div>
    </AccordionItem>
  )
}

it('should render title', () => {
  expect(
    prepareForTest()
      .find('.title')
      .get(0)
  ).toBeTruthy()
})

it('should always render title', () => {
  expect(
    prepareForTest({ show: false })
      .find('.title')
      .get(0)
  ).toBeTruthy()
})

it('should forward props as attrs', () => {
  expect(
    prepareForTest({ className: 'foo' })
      .find('.foo')
      .get(0)
  ).toBeTruthy()
})

it('should hide body content', () => {
  expect(
    prepareForTest()
      .find('.body')
      .get(0)
  ).toBeFalsy()
})

it('should show body content', () => {
  expect(
    prepareForTest({ show: true })
      .find('.body')
      .get(0)
  ).toBeTruthy()
})

it('should trigger onClick', () => {
  const spy = jest.fn()
  const wrapper = prepareForTest({ onClick: spy })

  wrapper.find('.title').simulate('click')

  expect(spy).toHaveBeenCalled()
})

it('should transition body exit', done => {
  const wrapper = prepareForTest({ show: true })

  expect(wrapper.find('.body').get(0)).toBeTruthy()

  wrapper.setProps({ show: false })

  expect(wrapper.find('.body').get(0)).toBeTruthy()

  setTimeout(() => {
    expect(wrapper.html()).not.toEqual(expect.stringContaining('class="body"'))
    expect(wrapper.html()).toMatchSnapshot()
    done()
  }, 360)
})
