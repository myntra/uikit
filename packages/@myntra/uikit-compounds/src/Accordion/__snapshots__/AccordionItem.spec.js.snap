// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should transition body exit 1`] = `"<div><div class=\\"title\\">Title</div></div>"`;
